cpp/src/io/parquet/page_enc.cu

```diff
@@ -1478,13 +1481,14 @@ __device__ uint8_t *EncodeStatistics(uint8_t *start,
 }
 
 // blockDim(128, 1, 1)
-__global__ void __launch_bounds__(128) gpuEncodePageHeaders(EncPage *pages,
-                                                            EncColumnChunk *chunks,
-                                                            const gpu_inflate_status_s *comp_out,
-                                                            const statistics_chunk *page_stats,
-                                                            const statistics_chunk *chunk_stats,
-                                                            uint32_t start_page)
+__global__ void __launch_bounds__(128)
+  gpuEncodePageHeaders(device_span<EncPage> pages,
```

Comment:
Since we are using a `subspan` here and still have `start_page`, the logic is a bit messy in this function. Some indexing is offset, some isn't. If we could completely eliminate `start_page` it would be :fire: . As it is now, I'm not a big fan of changes to this function.

Reply:
This is a side effect of using batched writing to keep the peak memory usage in check. Because of batches, the start page is different for each batch but it has to be indexed from the start of the pages array. A fix for this would've been to have a span of pages as a member of the chunk struct, so that a chunk knows the pages it contains. That posed a different problem because the chunk struct is loaded in shared memory in several places and shared memory doesn't allow default constructors that set member variables.

I'm not a big fan of this batches approach in general and plan to replace it all with a user controllable batch. This wasn't fully possible at the time I proposed it in the initial refactoring meetings but now that we have a per row size API in cudf, we should be able to use that and parquet chunked writer to give the user control full control.

Reply:
Finally did it! Still maintain my stance on batching within a single `write()` call, but that'll come later.